prow/spyglass/lenses/common/common.go

```diff
@@ -228,24 +235,42 @@ type ProwJobFetcher interface {
 
 // prowToGCS returns the GCS key corresponding to the given prow key
 // TODO: Unexport once we only have remote lenses
-func ProwToGCS(fetcher ProwJobFetcher, config config.Getter, prowKey string) (string, error) {
+func ProwToGCS(fetcher ProwJobFetcher, config config.Getter, prowKey string) (string, string, error) {
 	jobName, buildID, err := keyToJob(prowKey)
 	if err != nil {
-		return "", fmt.Errorf("could not get GCS src: %v", err)
+		return "", "", fmt.Errorf("could not get GCS src: %v", err)
 	}
 
 	job, err := fetcher.GetProwJob(jobName, buildID)
 	if err != nil {
-		return "", fmt.Errorf("Failed to get prow job from src %q: %v", prowKey, err)
+		return "", "", fmt.Errorf("failed to get prow job from src %q: %v", prowKey, err)
 	}
 
 	url := job.Status.URL
 	prefix := config().Plank.GetJobURLPrefix(job.Spec.Refs)
 	if !strings.HasPrefix(url, prefix) {
-		return "", fmt.Errorf("unexpected job URL %q when finding GCS path: expected something starting with %q", url, prefix)
+		return "", "", fmt.Errorf("unexpected job URL %q when finding GCS path: expected something starting with %q", url, prefix)
+	}
+
+	// try to parse storageProvider from DecorationConfig.GCSConfiguration.Bucket
+	// if it doesn't work fallback to URL parsing
+	if job.Spec.DecorationConfig != nil && job.Spec.DecorationConfig.GCSConfiguration != nil {
+		storageProvider, _, err := prowv1.ParseBucketName(job.Spec.DecorationConfig.GCSConfiguration.Bucket)
```

Comment:
In this case it's not about the bucket name itself or behaving differently depending on the bucket name. My only requirement is that I have to be able to build an URL with a storage provider prefix based on a prowjob URL. As far as I know the "prowjob" URLs look like this: `https://prow.k8s.io/view/prowjob/<prowjob-name>/1248318192296660992` (I couldn't find a place where this URLs are built nor did I find any examples of it, but the code looks like it handles these kind of URLs. Are these URLs still supported or documented anywhere?)

So If we still want to be able to handle these URLs (and also not only for GCS), I have to be able to find out the used StorageProvider somehow to decide If I should access GCS or S3.
common/common.go ~l.190:
```go
	case api.ProwKeyType:
		storageProvider, key, err := ProwToGCS(pjFetcher, cfg, key)
		if err != nil {
			logrus.Warningln(err)
		}
		gcsKey = fmt.Sprintf("%s://%s", storageProvider, strings.TrimSuffix(key, "/"))
...
	art, err := gcsArtifactFetcher.Artifact(ctx, gcsKey, name, sizeLimit)
```
`gcsArtifactFetcher` requires a path like `gs://..` or `s3://..` because otherwise we don't have a clue which storage provider to access.

I'm absolutely happy to change this handling if there's an alternative, I just don't know one.

Reply:
So the format is:
`HOST/view/<provider>/<job>/<build>`

where provider is either:
* `prowjob` -- aka access the kubernetes api directly, no storage configured
* `gcs` -- deprecated name for gs
* `other` -- uses the new opener to access it

Since we have GCS turned on for prow.k8s.io we try to never provide the prowjob URLs because they are ephemeral -- aka they stop working as soon as sinker deletes the pod and/or prowjob.

But you can create them:
* Go to deck and find a failing job - https://prow.k8s.io/?state=failure
  - Or any job, but sinker deletes these more slowly
* Click on the magnifying glass to get the job yaml - https://prow.k8s.io/prowjob?prowjob=fd653895-8ffc-11ea-990f-52c6b219fb85
  - note metadata.labels["prow.k8s.io/job"] = ci-benchmark-microbenchmarks
  - note status.build_id = 1258197944759226371
* Now construct that into https://prow.k8s.io/view/prowjob/ci-benchmark-microbenchmarks/1258197944759226371
  - Note this example will probably be broken by the time you read this..

Interestingly also look at status.url:
```yaml
  url: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-benchmark-microbenchmarks/1258197944759226371
```

So I suspect you can use this value... but maybe it is better to have deck detect that there is a url status field and autoredirect https://prow.k8s.io/view/prowjob/ci-benchmark-microbenchmarks/1258197944759226371 to https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-benchmark-microbenchmarks/1258197944759226371? And otherwise have view/prowjob only support reading information exclusively from what is found in the apiserver (nothing on storage).

<details>
<summary> yaml snippet from example </summary>

```yaml
metadata:
  annotations:
    prow.k8s.io/job: ci-benchmark-microbenchmarks
    testgrid-dashboards: sig-scalability-benchmarks
    testgrid-tab-name: microbenchmarks
  creationTimestamp: "2020-05-07T00:52:13Z"
  generation: 7
  labels:
    created-by-prow: "true"
    prow.k8s.io/job: ci-benchmark-microbenchmarks
    prow.k8s.io/type: periodic
  name: fd653895-8ffc-11ea-990f-52c6b219fb85
  namespace: default
  resourceVersion: "419795230"
  selfLink: /apis/prow.k8s.io/v1/namespaces/default/prowjobs/fd653895-8ffc-11ea-990f-52c6b219fb85
  uid: fd833ffe-8ffc-11ea-94be-42010a8000c3
spec:
  agent: kubernetes
  cluster: default
  decoration_config:
    gcs_configuration:
      bucket: kubernetes-jenkins
      default_org: kubernetes
      default_repo: kubernetes
      path_strategy: legacy
    gcs_credentials_secret: service-account
    grace_period: 15s
    timeout: 2h0m0s
    utility_images:
      clonerefs: gcr.io/k8s-prow/clonerefs:v20200501-e6124e633
      entrypoint: gcr.io/k8s-prow/entrypoint:v20200501-e6124e633
      initupload: gcr.io/k8s-prow/initupload:v20200501-e6124e633
      sidecar: gcr.io/k8s-prow/sidecar:v20200501-e6124e633
  extra_refs:
  - base_ref: master
    org: kubernetes
    path_alias: k8s.io/test-infra
    repo: test-infra
  - base_ref: master
    org: kubernetes
    path_alias: k8s.io/kubernetes
    repo: kubernetes
  job: ci-benchmark-microbenchmarks
  namespace: test-pods
  pod_spec:
    containers:
    - args:
      - --log-file=$(ARTIFACTS)/benchmark-log.txt
      - --output=$(ARTIFACTS)/junit_benchmarks.xml
      - --pass-on-error
      - ../kubernetes/pkg/...
      - ../kubernetes/plugin/...
      - ../kubernetes/vendor/k8s.io/apimachinery/...
      - ../kubernetes/vendor/k8s.io/apiserver/...
      - ../kubernetes/vendor/k8s.io/client-go/...
      command:
      - /benchmarkjunit
      env:
      - name: GOPROXY
        value: https://proxy.golang.org
      image: gcr.io/k8s-testimages/benchmarkjunit:latest
      name: ""
      resources: {}
  report: true
  type: periodic
status:
  build_id: "1258197944759226371"
  completionTime: "2020-05-07T00:54:50Z"
  description: Job failed.
  pendingTime: "2020-05-07T00:52:20Z"
  pod_name: fd653895-8ffc-11ea-990f-52c6b219fb85
  prev_report_states:
    gcsk8sreporter: failure
    gcsreporter: failure
    slackreporter: failure
  startTime: "2020-05-07T00:52:13Z"
  state: failure
  url: https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-benchmark-microbenchmarks/1258197944759226371
```

</summary>

Reply:
Thx for these infos :)
I'll start with writing a unit test for this based on your example (and I'll run deck locally with the config for prow.k8s.io to double check this). Let's see what the exact input/outputs of these func are and then what's the best implementation variant is

Reply:
Hmm okay so independent if I add a redirect or read directly from storage on the request to e.g. https://prow.k8s.io/view/prowjob/ci-benchmark-microbenchmarks/1258197944759226371 the problem remains the same that I have to either:
* redirect to https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-benchmark-microbenchmarks/1258197944759226371? or
* read from gs://kubernetes-jenkins/logs/ci-benchmark-microbenchmarks/1258197944759226371?

In both cases I have to find out which storageProvider should be:
* added to the URL or 
* used to read from storage

We can get this info from the ProwJob or the configured `job_url_prefix_config`, thus we have the options
1. get it from the ProwJob `status.url` or
2. get it from the ProwJob `DecorationConfig`
3. parse it from the last segment of the `job_url_prefix_config`

Currently I first try 2. and if this doesn't work 3. The problem with 1. is how to detect the correct URL segment for the storageProvider (`https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-benchmark-microbenchmarks/1258197944759226371`). I'm not sure if there is a way to detect the storageProvider correctly from only the `status.url`

P.S. I pushed the unit test for `ProwToGCS`
P.S.2 I just saw that as of now as long as `job_url_prefix_config` has to end with the storageProvider (`gcs`, `gs` or `s3`). Thus, one Prow instance won't be able to handle s3 and gs at the same time. I just fear if I also start to introduce changes to the `job_url_prefix_config` in this PR, it will make it a lot more likely to break something. But I think we need a follow-up PR with a refactoring which makes it possible to configure `job_url_prefix_config` with e.g. `https://prow.k8s.io/view/` and deprecate the format with storageProvider (`https://prow.k8s.io/view/gcs/`)